### ISSUE TITLE
Mechanism for getting from metadata symbol to underlying handle or raw metadata token

### ISSUE THREAD

**Issue by nguerrera** [dotnet/roslyn#6749]
It would be useful to be able to correlate a Roslyn symbol originating in metadata to PDBs, IL, and lower level metadata.

I've experimented with it and I'll submit the following as a straw man proposal: https://github.com/dotnet/roslyn/compare/master...nguerrera:metadata-symbol

``` C#
using System.Reflection.Metadata;

public interface IMetadataSymbol
{
    Handle MetadataHandle { get; }
}

internal abstract class PENamedTypeSymbol : NamedTypeSymbol, IMetadataSymbol
{
   public Handle MetadataHandle => _handle;
   ..
}

// etc. for all of the PEXxxSymbols for C# and VB.
```

This only adds a single new public type with one property, and the usage would involve querying for `IMetadataSymbol` dynamically, but that seems fine given that it is not a mainline scenario. Alternatively, it could be exposed for all symbols with non-metadata symbols returning a nil handle.

Another point to consider is whether it is OK to use a public dependency on `System.Reflection.Metadata` here or if the API should return a raw metadata token as an integer. The strongly-typing as Handle seems nice and it is always possible to drop down further using `System.Reflection.Metadata.MetadataTokens.GetToken`, but I'm not sure about the public API dependency.

cc @tmat

**Comment by tmat** [comment]
I think using Handle is preferable. This API would be used for very specific purpose which won't be that common. Note we already have an API that exposes MethodHandle for another specific scenario - EnC.

**Comment by nguerrera** [comment]
I've assigned this to myself and I'll update my branch along whatever unfolds as the design from the discussion here and contribute a PR when the design lands.

@tmat I didn't realize there was already such a dependency. In that case, it's pretty clear we should use Handle and not int. I also wondered if it should be EntityHandle, but that would preclude providing a NamespaceDefinitionHandle from namespace symbols. Today, we don't use that to build namespace symbols, so we wouldn't provide it initially, but it seems wise to allow for wider handles in the future.

What do you think of the IMetadataSymbol approach and definition above? Who else should I tag to look at it?

**Comment by tmat** [comment]
Re: namespaces -- I think we should move Roslyn over to using namespace handles. Currently it does namespace tree building itself, but it would be good to leverage the impl in the reader (unless we find some reason why it can't be done).

Re: the design of IMetadataSymbol -- @JohnHamby @mattwar @gafter @AnthonyDGreen @dotnet/roslyn-compiler

**Comment by AlekseyTs** [comment]
VB namespaces are case-insensitive.

**Comment by nguerrera** [comment]
> VB namespace are case-insensitive

Good point. Right now, that is enough to prevent a VB.PENamespaceSymbol from being backed by a single NamespaceDefinitionHandle. 

I still think using the base Handle (with the extra byte vs. EntityHandle) is most future-proof -- we could opt to use metadata reader namespace tree for C# only or even have case-insensitive handles to the reader (these handles are synthesized and we could build case-sensitive and insensitive trees on demand there too).

**Comment by tmat** [comment]
@nguerrera Couldn't we support case-insensitive namespaces  in the reader?

**Comment by nguerrera** [comment]
@tmat That's what I meant by "right now". We could add case-insensitive tree and handles if we want. It might clutter the S.R.M API (I'm not sure), but it is certainly possible.

**Comment by nguerrera** [comment]
cc @michaelcfanning

**Comment by nguerrera** [comment]
@JohnHamby @mattwar @gafter @AnthonyDGreen @dotnet/roslyn-compiler Any thoughts on this? Should I go ahead and clean up what I have with tests and so forth and submit a PR?

**Comment by nguerrera** [comment]
I've updated the proposal with a second option that I like better.

**Comment by agocke** [comment]
I do not like the second option where `ISymbol` has an added handle. Interfaces should contain only methods which are semantically meaningful for the interface. If a method is not semantically valid for an interface it should be split out to a new interface -- in this case `IMetadataSymbol`.

I think the `IMetadataSymbol` proposal is fine.

**Comment by mattwar** [comment]
I like the proposal with the separate interface too.

**Comment by JohnHamby** [comment]
My instinct is to prefer the separate interfaces approach, because the metadata handles are not of general interest and will look like clutter to most consumers.

However, maybe my instincts are wrong. I would like to hear a defense of the claim that adding metadata properties to the existing interfaces "is much better and cleaner".

**Comment by nguerrera** [comment]
I'm happy to go with the consensus on this since I don't feel overly strong about it, but I do still lean towards amending the existing interface. Here's some elaboration on why:

* I think it is sufficiently "semantically meaningful" on ISymbol. The semantics are clear: some symbols don't have metadata handles, which will result in a nil handle being returned. To me, it isn't all that different from say ISymbol.IsSealed, which is not a concept I'd associate with namespace, labels, or locals, but it still works because there is a reasonable default answer.

* It would avoid casting (slight perf gain).

* Some instances of symbols from metadata won't have handles: e.g. unnamed parameters. So with the separate interface, I'd have to do two checks whenever getting a handle: (1) query the interface; (2) validate that I didn't get nil. I think it would be easier to understand and get right if there was only the nil check involved.

* IAssemblySymbol and IModuleSymbol now have GetMetadata() members in this release. It seems that if we want to move metadata-only concerns to separate interfaces, then we might want to revisit that too. Otherwise, shouldn't we be consistent in the approach?

**Comment by gafter** [comment]
@nguerrera If this is still of interest to you, I will be happy to shepherd a PR into the compilers.

**Comment by nguerrera** [comment]
@gafter This is still interesting. The main work that needed it (IL-to-IOperation) hasn't got out of the prototype phase and for that I just used a private build of Roslyn to unblock that. The priority isn't urgent, but it is certainly interesting. By "shepherding" are you offering to review a PR if I submit one, or take over from here and submit the actual PR with tests and all? I'm hoping the latter. :grin: 

Also worth noting that #6748 has the same interest/priority for me and its need will arise in the same set of use cases as this. Neither is must-have for 2.0 (at least not for me).

**Comment by gafter** [comment]
@nguerrera I was offering to review a PR, and bug others to get it reviewed, and then integrate it. I'd rather not write the tests myself.

**Comment by nguerrera** [comment]
@gafter Ok. Thanks.
